tour_model_eval/evaluation_pipeline.py

```diff
@@ -0,0 +1,173 @@
+        # to the same cluster as each other.
+        method = 'single'
+        # get the second label from the second round of clustering using hierarchical clustering
+        second_labels = lp.get_second_labels(x, method, low, dist_pct)
+        # concatenate the first label (label from the first round) and the second label (label
+        # from the second round) (e.g.first label[1,1,1], second label[1,2,3], new_labels is [11,12,13]
+        new_labels = lp.get_new_labels(second_labels, second_round_idx_labels, new_labels)
+        # change the labels in track with new_labels
+        track = lp.change_track_labels(track, new_labels)
+
+    # get request percentage for the subset for the second round
+    percentage_second = grp.get_req_pct(new_labels, track, filter_trips, sim)
+
+    # get homogeneity score for the second round
+    homo_second = gs.score(bin_trips, new_labels)
+    return percentage_second,homo_second
+
+
+# we use functions in similarity to build the first round of clustering
+def first_round(data,radius):
+    sim = similarity.similarity(data, radius)
+    filter_trips = sim.data
+    sim.bin_data()
+    sim.delete_bins()
+    bins = sim.bins
+    bin_trips = sim.newdata
+    return sim, bins, bin_trips, filter_trips
+
+
+def get_first_label(bins):
+    # get first round labels
+    # the labels from the first round are the indices of bins
+    # e.g. in bin 0 [trip1, trip2, trip3], the labels of this bin is [0,0,0]
+    first_labels = []
+    for b in range(len(bins)):
+        for trip in bins[b]:
+            first_labels.append(b)
+    return first_labels
+
+
+def get_track(bins, first_labels):
+    # create a list idx_labels_track to store indices and labels
+    # the indices of the items will be the same in the new label list after the second round clustering
+    # item[0] is the original index of the trip in filter_trips
+    # item[1] is the label from the first round of clustering
+    idx_labels_track = []
+    for bin in bins:
+        for ori_idx in bin:
+            idx_labels_track.append([ori_idx])
+    # store first round labels in idx_labels_track list
+    for i in range(len(first_labels)):
+        idx_labels_track[i].append(first_labels[i])
+
+    return idx_labels_track
+
+
+def tune(data,radius):
+    sim, bins, bin_trips, filter_trips = first_round(data, radius)
+    # it is possible that we don't have common trips for tuning or testing
+    # bins contain common trips indices
+    if len(bins) is not 0:
+        gs.compare_trip_orders(bins, bin_trips, filter_trips)
+        first_labels = get_first_label(bins)
+        # new_labels temporary stores the labels from the first round, but later the labels in new_labels will be
+        # updated with the labels after two rounds of clustering.
+        new_labels = first_labels.copy()
+        first_label_set = list(set(first_labels))
+        track = get_track(bins, first_labels)
+        # collect tuning scores and parameters
```

Comment:
all of this is common, so it seems like it can be moved into `first_round` or into a new function like `get_first_round_labels` or something which is called from both `tune` and `test`

Reply:
There were just two lines of code repeated after refactoring, but I created a function         `get_first_label_and_track(bins,bin_trips,filter_trips)`, which can be called from both `tune` and `test`